### PR TITLE
Add lints for disallowing usage of `to_xx_bytes` and `from_xx_bytes`

### LINKED ISSUE #10765: Lint uses of to/from_ne_bytes
### What it does

Lint uses of `from_ne_bytes` and `to_ne_bytes` for code that needs to be endian aware. It may be desirable to force a developer to consider the endian which data should be stored in, especially when sending data across a network or to another machine.

### Lint Name

endian_bytes

### Category

restriction

### Advantage

- Useful for contexts where endianness needs to be considered (such as serialization of data).

### Drawbacks

Sometimes it may be desirable to use the host endianness. For example a protocol like Wayland uses host endian.

### Example

```rust
// 0xFE4415
let ferris = u32::from_ne_bytes([0x00, 0x15, 0x44, 0xFE]);
```

Could be written as:

```rust
// 0xFE4415
let ferris = u32::from_le_bytes([0x00, 0x15, 0x44, 0xFE]);
```

### PR DESCRIPTION
**Pull request by Centri3**
Adds `host_endian_bytes`, `little_endian_bytes` and `big_endian_bytes`

Closes #10765

v - not sure what to put here since this adds 3 lints
changelog: Add `host_endian_bytes`, `little_endian_bytes` and `big_endian_bytes` lints

### CODE REVIEW

**Review by Manishearth — APPROVED** [review]
This is a neat idea! I'm in favor, minor nit

File: clippy_lints/src/endian_bytes.rs [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+            if ty.is_primitive_ty();
+            then {
+                try_lint_endian_bytes(cx, expr, "from", *function_name, ty);
+            }
+        }
+    }
+}
+
+fn try_lint_endian_bytes(cx: &LateContext<'_>, expr: &Expr<'_>, prefix: &str, name: Symbol, ty: Ty<'_>) -> bool {
+    let ne = format!("{prefix}_ne_bytes");
```
[the remark this answers is not in the document]
Comment (Centri3) [reply clippy_lints/src/endian_bytes.rs]: Fixed

File: clippy_lints/src/endian_bytes.rs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+            let ty = cx.typeck_results().expr_ty(expr);
+            if ty.is_primitive_ty();
+            then {
+                maybe_lint_endian_bytes(cx, expr, "from", *function_name, ty);
+            }
+        }
+    }
+}
+
+fn maybe_lint_endian_bytes(cx: &LateContext<'_>, expr: &Expr<'_>, prefix: &str, name: Symbol, ty: Ty<'_>) -> bool {
```
Comment (Manishearth) [inline clippy_lints/src/endian_bytes.rs:150]: nit: I think `prefix` should be an enum or an `is_from` boolean

Comment (Centri3) [reply clippy_lints/src/endian_bytes.rs]: Fixed

File: clippy_lints/src/endian_bytes.rs [hunk 3]
```diff
@@ -211,9 +200,13 @@ fn maybe_lint_endian_bytes(cx: &LateContext<'_>, expr: &Expr<'_>, prefix: &str,
         expr.span,
         &format!(
             "usage of the {}`{ty}::{}`{}",
-            if prefix == "from" { "function " } else { "" },
-            lint.to_name(prefix),
-            if prefix == "to" { " method" } else { "" },
+            if matches!(prefix, Prefix::From) {
```
Comment (Manishearth) [inline clippy_lints/src/endian_bytes.rs:203]: nit: probably just use `==`, `derive(PartialEq)` the enum

File: clippy_lints/src/endian_bytes.rs [hunk 4]
```diff
@@ -127,7 +116,7 @@ impl LateLintPass<'_> for EndianBytes {
             if args.is_empty();
             let ty = cx.typeck_results().expr_ty(receiver);
             if ty.is_primitive_ty();
-            if maybe_lint_endian_bytes(cx, expr, "to", method_name.ident.name, ty);
+            if maybe_lint_endian_bytes(cx, expr, &Prefix::To, method_name.ident.name, ty);
```
Comment (Manishearth) [inline clippy_lints/src/endian_bytes.rs:119]: nit: `derive(Clone, Copy)` the enum and use the value directly instead of a reference

Comment (Centri3) [reply clippy_lints/src/endian_bytes.rs]: Fixed

File: clippy_lints/src/endian_bytes.rs (hunk 3 above)
Comment (Centri3) [reply clippy_lints/src/endian_bytes.rs]: Fixed

**Comment by Manishearth** [comment]
@bors r+

thanks!

**Pull request merged**